src/main/scala/io/kaizensolutions/virgil/internal/CQLExecutorImpl.scala

```diff
@@ -137,12 +139,22 @@ private[virgil] class CQLExecutorImpl(underlyingSession: CqlSession) extends CQL
   private def select(query: Statement[_]): ZStream[Any, Throwable, Row] = {
     val initialEffect = ZIO.fromCompletionStage(underlyingSession.executeAsync(query))
 
-    ZStream.paginateChunkM(initialEffect) { eff =>
-      eff.map { resultSet =>
-        val emit = Chunk.fromIterable(resultSet.currentPage().asScala)
-        val nextState = if (resultSet.hasMorePages) Option(ZIO.fromCompletionStage(resultSet.fetchNextPage())) else None
-        emit -> nextState
-      }
+    def pull(ref: Ref[ZIO[Any, Option[Throwable], AsyncResultSet]]): ZIO[Any, Option[Throwable], Chunk[Row]] =
+      for {
+        io <- ref.get
+        rs <- io
+        _ <- rs match {
+               case _ if rs.hasMorePages =>
+                 ref.set(Task.fromCompletionStage(rs.fetchNextPage()).mapError(Option(_)))
+               case _ if rs.currentPage().iterator().hasNext => ref.set(IO.fail(None))
+               case _                                        => IO.fail(None)
+             }
+      } yield Chunk.fromIterable(rs.currentPage().asScala)
```

Comment:
Just wanted to verify: 
1. there are more pages - update ref to retrieve next page on the next pull
2. there are no more pages but the current page has results - update ref to signal stop on the next pull
3. there are no more pages and the current page has no results - fail & stop immediately

For 2, you can use `rs.remaining > 0` which is built into the AsyncResultSet which corresponds to the number of rows are left before the current page is exhausted to avoid using the iterator.

Reply:
Yep. Agree with point :+1: